vowpalwabbit/cbify.cc

```diff
@@ -753,59 +753,95 @@ base_learner* cbify_setup(VW::setup_base_i& stack_builder)
   }
 
   learner<cbify, example>* l;
+  void (*finish_ptr)(vw&, cbify&, example&);
+  std::string name_addition;
+  label_type_t label_type;
 
   if (data->use_adf)
   {
+    void (*learn_ptr)(cbify&, multi_learner&, example&);
+    void (*predict_ptr)(cbify&, multi_learner&, example&);
     multi_learner* base = as_multiline(stack_builder.setup_base_learner());
 
     if (data->use_adf) { data->adf_data.init_adf_data(num_actions, base->increment, all.interactions); }
 
     if (use_cs)
     {
-      l = &init_cost_sensitive_learner(data, base, learn_adf<true>, predict_adf<true>, all.example_parser, 1,
-          stack_builder.get_setupfn_name(cbify_setup) + "-adf-cs");
-      all.example_parser->lbl_parser.label_type = label_type_t::cs;
+      learn_ptr = learn_adf<true>;
+      predict_ptr = predict_adf<true>;
+      finish_ptr = COST_SENSITIVE::finish_example;
+      name_addition = "-adf-cs";
+      label_type = label_type_t::cs;
+      all.example_parser->lbl_parser = COST_SENSITIVE::cs_label;
     }
     else
     {
-      l = &init_multiclass_learner(data, base, learn_adf<false>, predict_adf<false>, all.example_parser, 1,
-          stack_builder.get_setupfn_name(cbify_setup) + "-adf");
-      all.example_parser->lbl_parser.label_type = label_type_t::multiclass;
+      learn_ptr = learn_adf<false>;
+      predict_ptr = predict_adf<false>;
+      finish_ptr = MULTICLASS::finish_example<cbify&>;
+      name_addition = "-adf";
+      label_type = label_type_t::multiclass;
+      all.example_parser->lbl_parser = MULTICLASS::mc_label;
     }
+    l = make_reduction_learner(
+        std::move(data), base, learn_ptr, predict_ptr, stack_builder.get_setupfn_name(cbify_setup) + name_addition)
+            .set_finish_example(finish_ptr)
+            .set_label_type(label_type)
```

Comment:
What's the prediction type in this case?

Reply:
Looks like multiclass, I'll put that in